File: src/main/java/com/jeffreyorazulike/simpletron/calculator/InfixPostfix.java
/*
 * Copyright (C) 2021 Jeffrey Orazulike
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package com.jeffreyorazulike.simpletron.calculator;

import java.util.Stack;
import java.util.function.Function;
import java.util.function.Supplier;

/**
 *
 * @author Jeffrey Orazulike
 * @apiNote {@code InfixPostfix} helps the
 * {@link com.jeffreyorazulike.simpletron.translator.Translator} translate a
 * mathematical expression to a simpletron machine readable code
 */
public final class InfixPostfix {

    public static final char SPACE = ' ';
    public static final char PLUS = '+';
    public static final char MINUS = '-';
    public static final char DIVISION = '/';
    public static final char REMAINDER = '%';
    public static final char EXPONENTATION = '^';
    public static final char MULTIPLICATION = '*';
    public static final char O_PARENTHESIS = '(';
    public static final char C_PARENTHESIS = ')';

    @FunctionalInterface
    public static interface CalculateInterface {

        int calculate(Stack<Integer> stack, char operator, int count);
    }

    public static final class ReturnedNumber {

        public int number;
        public int i;

        public ReturnedNumber(int number, int i) {
            this.number = number;
            this.i = i;
        }
    }

    public static final CalculateInterface defaultCalculate = (stack, operator, count) -> {
        int y = stack.pop();
        int x = stack.pop();

        Supplier<Integer> calculate = () -> {
            switch (operator) {
                case PLUS:
                    return x + y;
                case MINUS:
                    return x - y;
                case MULTIPLICATION:
                    return x * y;
                case DIVISION:
                    return x / y;
                case REMAINDER:
                    return x % y;
                case EXPONENTATION:
                    return (int) Math.pow(x, y);

                default:
                    return 0;
            }
        };

        stack.push(calculate.get());
        return ++count;
    };

    /**
     *
     * @param start  the pos tion to start looking for a number
     * @param string the string to search
     *
     * @return the number and the current index
     */
    public static ReturnedNumber getNumbers(int start, String string) {
        StringBuilder number = new StringBuilder(string.length());
        char c;
        for (; start < string.length(); ++start) {
            c = string.charAt(start);
            if (Character.isDigit(c))
                number.append(c);
            else
                break;
        }
        return new ReturnedNumber(Integer.parseInt(number.toString()), start == string.length() ? start : start - 1);
    }

    /**
     *
     * @param operator The operator to check
     *
     * @return A number signifying precedence
     */
    private static int precedence(char operator) {
        switch (operator) {
            case PLUS:
            case MINUS:
                return 1;
            case DIVISION:
            case REMAINDER:
            case MULTIPLICATION:
                return 2;
            case EXPONENTATION:
                return 3;

            default:
                return 0;
        }
    }

    public static boolean isOperator(char c) {
        switch (c) {
            case PLUS:
            case MINUS:
            case DIVISION:
            case REMAINDER:
            case EXPONENTATION:
            case MULTIPLICATION:
                return true;
        }

        return false;
    }

    public static boolean isSpecial(char c) {
        if (isOperator(c))
            return true;
        switch (c) {
            case O_PARENTHESIS:
            case C_PARENTHESIS:
                return true;
        }
        return false;
    }

    public static String infixToPostfix(String infix) {
        StringBuilder postfix = new StringBuilder(infix.length());
        Stack<Character> stack = new Stack<>();
        stack.push(O_PARENTHESIS);
        infix = infix.concat(String.valueOf(C_PARENTHESIS));

        char c;
        for (int i = 0; i < infix.length(); ++i) {

            c = infix.charAt(i);

            if (Character.isWhitespace(c))
                continue;

            if (Character.isDigit(c)) {
                ReturnedNumber returnedNumber = getNumbers(i, infix);
                i = returnedNumber.i;
                postfix.append(returnedNumber.number).append(SPACE);
            } else if (Character.isAlphabetic(c))
                postfix.append(c).append(SPACE);
            else if (c == O_PARENTHESIS)
                stack.push(c);
            else if (isOperator(c)) {
                while (!stack.isEmpty() && (precedence(stack.peek()) >= precedence(c)))
                    postfix.append(stack.pop()).append(SPACE);
                stack.push(c);
            } else if (c == C_PARENTHESIS) {
                while (stack.peek() != O_PARENTHESIS)
                    postfix.append(stack.pop()).append(SPACE);
                stack.pop();
            }
        }

        return postfix.toString();
    }

    public static int postfixEvaluator(String postfix, CalculateInterface calculate, Function<Integer, Integer> digitTransform, Function<Character, Integer> variableTransform) {
        if (postfix.isBlank())
            return 0;

        final Stack<Integer> stack = new Stack<>();
        final String[] variables = postfix.concat(" ").concat(String.valueOf(C_PARENTHESIS)).split("\\s+");
        int count = 0;

        char c;
        for (String variable : variables) {
            c = variable.charAt(0);

            if (Character.isDigit(c))
                stack.push(digitTransform.apply(Integer.parseInt(variable)));
            else if (Character.isAlphabetic(c))
                stack.push(variableTransform.apply(c));
            else if (isOperator(c))
                count = calculate.calculate(stack, c, count);
            else if (c == C_PARENTHESIS && !stack.isEmpty())
                return stack.pop();
        }

        return 0;
    }

}

File: infix-postfix/src/main/kotlin/com/jeffreyorazulike/simpletron/compiler/infix_postfix/InfixPostfix.kt
package com.jeffreyorazulike.simpletron.compiler.infix_postfix

import java.util.*
import kotlin.math.pow

const val DECIMAL = '.'
const val SPACE = ' '
const val PLUS = '+'
const val MINUS = '-'
const val DIVISION = '/'
const val REMAINDER = '%'
const val EXPONENTIATION = '^'
const val MULTIPLICATION = '*'
const val O_PARENTHESIS = '('
const val C_PARENTHESIS = ')'

private val defaultCalculate: (Stack<Float>, Char) -> Float = { stack, operator ->
    val y = stack.pop()
    val x = stack.pop()

    when (operator) {
        PLUS -> x + y
        MINUS -> x - y
        MULTIPLICATION -> x * y
        DIVISION -> x / y
        REMAINDER -> x % y
        EXPONENTIATION -> x.toDouble().pow(y.toDouble()).toFloat()
        else -> 0f
    }
}

private val defaultTransform: (String) -> Float = { string ->
    when (string.length) {
        1 -> {
            val char = string.toCharArray()[0]
            if (char.isDigit()) string.toFloat() else char.code.toFloat()
        }
        else -> string.toFloat()
    }
}

/**
 * Returns true if this character is an operator, false otherwise
 * */
fun Char.isOperator(): Boolean {
    return when (this) {
        PLUS,
        MINUS,
        DIVISION,
        REMAINDER,
        EXPONENTIATION,
        MULTIPLICATION -> true

        else -> false
    }
}

/**
 *
 * Returns a whole number from the start index to the index of a non number character
 *
 * @param start the position to start looking for a number
 */
fun CharArray.getNumber(start: Int): Pair<Number, Int> {
    val number = StringBuilder(size)
    var i = start
    while (i < size) {
        val c = this[i]
        if (c.isDigit() || c == DECIMAL) {
            number.append(c)
            ++i
        } else {
            --i
            break
        }
    }

    return if (number.contains(DECIMAL))
        number.toString().toFloat() to i
    else
        number.toString().toInt() to i
}

/**
 * Returns thr precedence of a character if it is an operator or zero if it is not
 */
fun Char.precedence(): Int {
    return when (this) {
        PLUS, MINUS -> 1
        DIVISION, REMAINDER, MULTIPLICATION -> 2
        EXPONENTIATION -> 3
        else -> 0
    }
}

/**
 * Converts an infix expression to postfix
 * */
fun String.infixToPostfix(): String {
    val chars = plus(C_PARENTHESIS).replace(Regex("\\s+"), "").toCharArray().also {
        if (it.isEmpty()) return ""
    }
    val postfix = StringBuilder(chars.size)
    val stack = Stack<Char>().apply { push(O_PARENTHESIS) }

    var i = 0
    do {
        val char = chars[i]
        when {
            char.isDigit() || char == DECIMAL -> chars.getNumber(i).apply {
                postfix.append(this.first).append(SPACE)
                i = this.second
            }
            char.isLetter() -> postfix.append(char).append(SPACE)
            char == O_PARENTHESIS -> stack.push(char)
            char.isOperator() -> {
                while (stack.isNotEmpty() && stack.peek().precedence() >= char.precedence())
                    postfix.append(stack.pop()).append(SPACE)
                stack.push(char)
            }
            char == C_PARENTHESIS -> {
                while (stack.peek() != O_PARENTHESIS)
                    postfix.append(stack.pop()).append(SPACE)
                stack.pop()
            }
        }
    } while (++i < chars.size)
    return postfix.toString().trim()
}

/**
 * Evaluates a postfix expression
 *
 * @param transform changes a letter of digit to it corresponding int value
 * @param calculate evaluates a simple expression between two int values
 * */
fun String.postfixEvaluator(
    transform: (String) -> Float = defaultTransform,
    calculate: (Stack<Float>, Char) -> Float = defaultCalculate
): Number {
    val stack = Stack<Float>()
    val variables = plus(" $C_PARENTHESIS").split(Regex("\\s+"))

    for (variable in variables) {
        val char = variable[0]
        when {
            char == C_PARENTHESIS -> {
                val value = stack.pop()
                val valueInt = value.toInt()
                // if the value after the decimal point of the float is only zero
                return if (valueInt.toFloat() == value)
                // return the integer version
                    valueInt
                else
                // return the float version stopping at the 100th value
                    "%.2f".format(value).toFloat()
            }
            char.isLetterOrDigit() -> stack.push(transform(variable))
            char.isOperator() -> stack.push(calculate(stack, char))
        }
    }
    return 0
}